File: src/lib/api/targets.ts
import type {
  Target,
  CreateTargetParams,
  UpdateTargetParams,
} from "@/lib/types";
import { invoke, withTauriError } from "./utils";

export async function getTarget(id: string): Promise<Target> {
  return withTauriError("获取 Target", async () => {
    return invoke<Target>("get_target", { id });
  });
}

export async function getTargets(): Promise<Target[]> {
  return withTauriError("获取 Target 列表", async () => {
    return invoke<Target[]>("get_targets");
  });
}

export async function createTarget(data: CreateTargetParams): Promise<Target> {
  return withTauriError("创建 Target", async () => {
    return invoke<Target>("create_target", {
      title: data.title,
      description: data.description || null,
      due_date: data.due_date || null,
    });
  });
}

export async function updateTarget(
  id: string,
  data: UpdateTargetParams,
): Promise<Target> {
  return withTauriError("更新 Target", async () => {
    return invoke<Target>("update_target", {
      id,
      title: data.title,
      description: data.description,
      due_date: data.due_date,
      status: data.status,
    });
  });
}

export async function deleteTarget(id: string): Promise<void> {
  return withTauriError("删除 Target", async () => {
    return invoke<void>("delete_target", { id });
  });
}

File: src-tauri/src/commands/targets.rs
// Target CRUD commands

use crate::commands::repositories::TargetRepository;
use crate::log_command;
use crate::models::Target;
use crate::AppState;

#[tauri::command]
pub fn get_target(state: tauri::State<AppState>, id: String) -> Result<Target, String> {
    log_command!("get_target", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TargetRepository::get_by_id(&conn, &id)
    })
}

#[tauri::command]
pub fn get_targets(state: tauri::State<AppState>) -> Result<Vec<Target>, String> {
    log_command!("get_targets", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TargetRepository::get_all(&conn)
    })
}

#[tauri::command]
pub fn create_target(
    state: tauri::State<AppState>,
    title: String,
    description: Option<String>,
    due_date: Option<String>,
) -> Result<Target, String> {
    log_command!("create_target", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        let id = uuid::Uuid::new_v4().to_string();
        TargetRepository::create(
            &conn,
            &id,
            &title,
            description.as_deref(),
            due_date.as_deref(),
        )
    })
}

#[tauri::command]
pub fn update_target(
    state: tauri::State<AppState>,
    id: String,
    title: Option<String>,
    description: Option<String>,
    due_date: Option<String>,
    status: Option<String>,
) -> Result<Target, String> {
    log_command!("update_target", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TargetRepository::update(
            &conn,
            &id,
            title.as_deref(),
            description.as_deref(),
            due_date.as_deref(),
            status.as_deref(),
        )
    })
}

#[tauri::command]
pub fn delete_target(state: tauri::State<AppState>, id: String) -> Result<(), String> {
    log_command!("delete_target", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TargetRepository::delete(&conn, &id)
    })
}

#[tauri::command]
pub fn update_target_sort_order(
    state: tauri::State<AppState>,
    id: String,
    sort_order: i32,
) -> Result<(), String> {
    log_command!("update_target_sort_order", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TargetRepository::update_sort_order(&conn, &id, sort_order)
    })
}

#[tauri::command]
pub fn reorder_targets(
    state: tauri::State<AppState>,
    orders: Vec<(String, i32)>,
) -> Result<usize, String> {
    log_command!("reorder_targets", {
        let conn = state.db.lock().map_err(|e| e.to_string())?;
        TargetRepository::reorder(&conn, &orders)
    })
}
